components/use-paid-mutation.js

```diff
@@ -0,0 +1,165 @@
+import { useApolloClient, useMutation } from '@apollo/client'
+import { useCallback, useState } from 'react'
+import { InvoiceCanceledError, InvoiceExpiredError, useInvoice, useQrPayment, useWebLnPayment } from './payment'
+
+/*
+this is just like useMutation with a few changes:
+1. pays an invoice returned by the mutation
+2. takes an onPaid and onPayError callback, and additional options for payment behavior
+  - namely forceWaitForPayment which will always wait for the invoice to be paid
+  - and persistOnNavigate which will keep the invoice in the cache after navigation
+3. onCompleted behaves a little differently, but analogously to useMutation, ie clientside side effects
+  of completion can still rely on it
+  a. it's called before the invoice is paid for optimistic updates
+  b. it's called after the invoice is paid for pessimistic updates
+4. we return a payError field in the result object if the invoice fails to pay
+*/
+export function usePaidMutation (mutation,
+  { onCompleted, ...options } = {}) {
+  options.optimisticResponse = addOptimisticResponseExtras(options.optimisticResponse)
+  const [mutate, result] = useMutation(mutation, options)
+  const waitForWebLnPayment = useWebLnPayment()
+  const waitForQrPayment = useQrPayment()
+  const invoiceWaiter = useInvoice()
+  const client = useApolloClient()
+  // innerResult is used to store/control the result of the mutation when innerMutate runs
+  const [innerResult, setInnerResult] = useState(result)
```

Comment:
Is this needed anywhere? I searched for `usePaidMutation` and I only found code which only used the first returned element.

Reply:
I'm not sure what you mean exactly. Do you mean is innerResult needed? We return it from the function.

The idea with `usePaidMutation` to mostly mirror `useMutation` so that if you're familiar with calling `useMutation`, it should be easy to use `usePaidMutation`.

Reply:
> We return it from the function.

Yes but I don't see it used anywhere and I wondered if that was just old code or something.

> The idea with `usePaidMutation` to mostly mirror `useMutation` so that if you're familiar with calling `useMutation`, it should be easy to use `usePaidMutation`.

Ah, right, that makes sense.